odbc/src/api/functions.rs

```diff
@@ -1370,20 +1370,31 @@ pub unsafe extern "C" fn SQLFetch(statement_handle: HStmt) -> SqlReturn {
             match res {
                 Err(e) => {
                     stmt.errors.write().unwrap().push(e.into());
-                    return SqlReturn::ERROR;
+                    SqlReturn::ERROR
                 }
-                Ok(b) => {
+                Ok((b, e)) => {
                     let mut stmt_attrs = stmt.attributes.write().unwrap();
+                    if let Some(errors) = e.clone() {
```

Comment:
I was hoping to do this when I initially tried, but calling the macro was complaining about borrowing rules, since the `mongo_handle` has already been used immutably (to get the stmt I believe). I'm not sure if this was intentional, but noticed this in a couple other places in the `SQLFetch` (e.g. [here](https://github.com/mongodb/mongo-odbc-driver/blob/11461a78c549d3afa1e808e0bc11c63b190b1d99/odbc/src/api/functions.rs#L1372)).

One way to work around this would be to make _new_ references to the handle, such as ```add_diag_with_function!(MongoHandleRef::from(statement_handle), warning, "SQLFetch");```, but borrowing the same handle multiple times feels unsafe to me. I'm still getting used to the concept of borrowing however, so definitely possible I'm missing something.

Reply:
Here is a snippet for SQLFetch which adds the Warnings to the diagnostic records for the handle and log them.
It first populates the diag. records, even though if the result is SQL_NO_DATA I bet now now will call SQLGetDiagRec, but warnings will be logged in every case this way.
If you remove the Option on the Warnings, just check the length of the vector.
```
pub unsafe extern "C" fn SQLFetch(statement_handle: HStmt) -> SqlReturn {
    panic_safe_exec!(
        || {
            let mongo_handle = MongoHandleRef::from(statement_handle);
            let stmt = must_be_valid!(mongo_handle.as_statement());
            let move_to_next_result = {
                let connection = must_be_valid!((*stmt.connection).as_connection());
                match stmt.mongo_statement.write().unwrap().as_mut() {
                    Some(mongo_stmt) => mongo_stmt
                        .next(connection.mongo_connection.read().unwrap().as_ref())
                        .map_err(|e| e.into()),
                    None => Err(ODBCError::InvalidCursorState),
                }
            };

            if let Ok((has_next, warnings_opt)) = move_to_next_result {
                let mut stmt_attrs = stmt.attributes.write().unwrap();

                // Add the warning to the diagnostic records and log them
                let has_warnings = warnings_opt.is_some();
                if let Some(warnings) = warnings_opt {
                    warnings.iter().for_each(|warning| {
                        add_diag_info!(MongoHandleRef::from(statement_handle), warning.into());
                    });
                }
                if !has_next {
                    stmt_attrs.row_index_is_valid = false;
                    // No more rows
                    return SqlReturn::NO_DATA;
                }
                stmt_attrs.row_index_is_valid = true;

                *stmt.var_data_cache.write().unwrap() = Some(HashMap::new());

                if has_warnings {
                    // No warnings and there is a next row
                    SqlReturn::SUCCESS_WITH_INFO
                } else {
                    SqlReturn::SUCCESS
                }
            } else {
                add_diag_info!(
                    mongo_handle,
                    move_to_next_result.as_ref().unwrap_err().clone()
                );
                // An error happened
                SqlReturn::ERROR
            }
        },
        statement_handle
    );
}

```